onboarding/source/amb_light_service.c

```diff
@@ -18,22 +25,63 @@ static void lightServiceTask(void * pvParameters);
 uint8_t initLightService(void) {
     /* USER CODE BEGIN */
     // Create the task and queue here.
+    if (lightServiceTaskHandle == NULL) {
+        BaseType_t xReturned = xTaskCreate(lightServiceTask,
+                                            LIGHT_SERVICE_NAME,
+                                            LIGHT_SERVICE_STACK_SIZE,
+                                            NULL,
+                                            LIGHT_SERVICE_PRIORITY,
+                                            &lightServiceTaskHandle);
+        if (xReturned == pdFAIL) {
+            sciPrintText(scilinREG, (unsigned char *) "Error: Cannot start light service task", 38);
+            return 0;
+        }
+    }
 
+    if (lightQueueHandle == NULL) {
+        lightQueueHandle = xQueueCreate(QUEUE_SIZE, QUEUE_ITEM_SIZE);
+        if (lightQueueHandle == NULL) {
+            sciPrintText(scilinREG, (unsigned char *) "Error: Cannot initialize light queue", 36);
+            return 0;
+        }
+    }
     /* USER CODE END */
     return 1;
 }
 
 static void lightServiceTask(void * pvParameters) {
     /* USER CODE BEGIN */
     // Wait for MEASURE_LIGHT event in the queue and then print the ambient light value to the serial port.
+    if (lightQueueHandle != NULL) {
+        light_event_t event;
+        if (xQueueReceive(lightQueueHandle, &event, 1000) == pdPASS) {
+            if (event == MEASURE_LIGHT) {
+                adcData_t data;
+                adcData_t * dataPtr = &data;
+                adcStartConversion(adcREG1, adcGROUP1);
+                while (!adcIsConversionComplete(adcREG1, adcGROUP1));
+                adcGetData(adcREG1, adcGROUP1, dataPtr);
 
+                uint16 value = dataPtr->value;
+                uint32_t outputLength = (uint32_t) log10(value);
+                char output[outputLength + 1];
+                snprintf(output, outputLength, "%d", value);
+                sciPrintText(scilinREG, (unsigned char *) output, outputLength);
+            }
+        }
+    }
     /* USER CODE END */
 }
 
 uint8_t sendToLightServiceQueue(light_event_t *event) {
     /* USER CODE BEGIN */
     // Send the event to the queue.
-    
+    if (lightQueueHandle != NULL) {
+        if (xQueueSend(lightQueueHandle, MEASURE_LIGHT, 1000) != pdPASS) {
```

Comment:
Even though `MEASURE_LIGHT` is the only light event we have so far, it would make sense for this to work for any light event, in case we add more in the future.

Reply:
To add to this, the `xQueueSend` function expects a pointer to the queue item. If you pass in `MEASURE_LIGHT` directly, what'll actually happen is it'll treat that as an address. 

Since `MEASURE_LIGHT` = 0, the program will think the queue item is stored at address 0 which isn't the case. If you try to dereference 0, you'll get a segmentation fault.

As a side note: you never want to implicitly convert between a pointer and integer (like `int *p = 5;`). Most compilers will say that's an error. If you ever need to set a pointer to a specific memory address (like address 0x3011FFE4), then you should do `int *p = (int *) 0x3011FFE4;`.